File: cflib_2_12/cplus/libcflib/CFLibArgumentUnderflowException.cpp
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright (c) 2020 Mark Stephen Sobkow
 *	
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *	
 *	    http://www.apache.org/licenses/LICENSE-2.0
 *	
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

#include <cflib/CFLib.hpp>
#include <cflib/CFLibXmlUtil.hpp>
#include <cflib/CFLibArgumentUnderflowException.hpp>

namespace cflib {

	const std::string CFLibArgumentUnderflowException::S_MUST_BE_AT_LEAST( " must be at least " );

	CFLibArgumentUnderflowException::~CFLibArgumentUnderflowException() {
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const CFLibArgumentUnderflowException& src )
	: std::underflow_error( src )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& msg )
	: std::underflow_error( msg )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		const std::string& msg )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE + msg )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		const std::string& msg )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + msg )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		int16_t argValue,
		int16_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		int32_t argValue,
		int32_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		int64_t argValue,
		int64_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		uint16_t argValue,
		uint16_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		uint32_t argValue,
		uint32_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		uint64_t argValue,
		uint64_t minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		float argValue,
		float minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		double argValue,
		double minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + std::to_string( argValue ) + S_MUST_BE_AT_LEAST + std::to_string( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		const std::chrono::system_clock::time_point& argValue,
		const std::chrono::system_clock::time_point& minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + CFLibXmlUtil::formatTimestamp( argValue ) + S_MUST_BE_AT_LEAST + CFLibXmlUtil::formatTimestamp( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		const std::string& argValue,
		const std::string& minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + argValue + S_MUST_BE_AT_LEAST + minValue )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		mpfr_srcptr argValue,
		mpfr_srcptr minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + CFLibXmlUtil::formatMPFR( argValue ) + S_MUST_BE_AT_LEAST + CFLibXmlUtil::formatMPFR( minValue ) )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& throwingClassName,
		const std::string& methName,
		int argNo,
		const std::string& argName,
		const CFLibGenericBigDecimal& argValue,
		const CFLibGenericBigDecimal& minValue )
	: std::underflow_error( throwingClassName + CFLib::S_DOUBLE_COLON + methName + CFLib::S_PAREN_SPACE_ARGUMENT + std::to_string( argNo ) + CFLib::S_SPACE + argName + CFLib::S_SPACE + CFLib::S_VALUE_SPACE + argValue.toString() + S_MUST_BE_AT_LEAST + minValue.toString() )
	{
	}

	CFLibArgumentUnderflowException::CFLibArgumentUnderflowException(
		const std::string& fieldName,
		const std::string& msg )
	: std::underflow_error( fieldName + CFLib::S_SPACE + msg )
	{
	}
}

File: cflib_2_12/csharp/CFLib/CFLibArgumentUnderflowException.cs
/*
 *  MSS Code Factory CFLib 2.11
 *
 *	Copyright (c) 2019 Mark Stephen Sobkow
 *	
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *	
 *		http://www.apache.org/licenses/LICENSE-2.0
 *	
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

using System;
using System.Globalization;

namespace CFLib
{
	public class CFLibArgumentUnderflowException : CFLibArgumentException {

		protected Object objValue = null;
		protected Object objMin = null;

		public CFLibArgumentUnderflowException(
			String msg )
		: base(msg )
		{
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			String msg )
		: base(throwingType, methName, msg)
		{
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			String msg,
			Exception th )
		: base(throwingType, methName, msg, th)
		{
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			String msg )
		: base(throwingType, methName, argNo, argName, msg)
		{
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			String msg,
			Exception th )
		: base(throwingType, methName, argNo, argName, msg, th)
		{
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			short argValue,
			short maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt16(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt16(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			short argValue,
			short maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt16(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt16(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			int argValue,
			int maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt32(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt32(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			int argValue,
			int maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt32(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt32(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			long argValue,
			long maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt64(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt64(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			long argValue,
			long maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt64(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt64(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			float argValue,
			float maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatSingle(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatSingle(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			float argValue,
			float maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatSingle(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatSingle(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			double argValue,
			double maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatDouble(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatDouble(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			double argValue,
			double maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatDouble(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatDouble(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			DateTime argValue,
			DateTime maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			DateTime argValue,
			DateTime maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			String argValue,
			String maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			String argValue,
			String maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			BigDecimal argValue,
			BigDecimal maxValue )
		: base(throwingType, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			Type throwingType,
			String methName,
			int argNo,
			String argName,
			BigDecimal argValue,
			BigDecimal maxValue,
			Exception th )
		: base(throwingType, methName, argNo, argName,
					"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String msg )
		: base(fieldName, msg)
		{
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			String msg )
		: base(fieldName, methName, msg)
		{
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			String msg,
			Exception th )
		: base(fieldName, methName, msg, th)
		{
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			String msg )
		: base(fieldName, methName, argNo, argName, msg)
		{
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			String msg,
			Exception th )
		: base(fieldName, methName, argNo, argName, msg, th)
		{
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			short argValue,
			short maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt16(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt16(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			short argValue,
			short maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt16(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt16(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			int argValue,
			int maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt32(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt32(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			int argValue,
			int maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt32(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt32(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			long argValue,
			long maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt64(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt64(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			long argValue,
			long maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatInt64(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatInt64(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			float argValue,
			float maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatSingle(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatSingle(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			float argValue,
			float maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatSingle(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatSingle(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			double argValue,
			double maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatDouble(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatDouble(maxValue ) )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			double argValue,
			double maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + CFLibXmlUtil.formatDouble(argValue )
					+ " must be at least "
					+ CFLibXmlUtil.formatDouble(maxValue ),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			DateTime argValue,
			DateTime maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			DateTime argValue,
			DateTime maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			String argValue,
			String maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			String argValue,
			String maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			BigDecimal argValue,
			BigDecimal maxValue )
		: base(fieldName, methName, argNo, argName,
				"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString() )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public CFLibArgumentUnderflowException(
			String fieldName,
			String methName,
			int argNo,
			String argName,
			BigDecimal argValue,
			BigDecimal maxValue,
			Exception th )
		: base(fieldName, methName, argNo, argName,
					"value " + argValue.ToString()
					+ " must be at least "
					+ maxValue.ToString(),
				th )
		{
			objValue = argValue;
			objMin = maxValue;
		}

		public Object getValue() {
			return( objValue );
		}
		
		public Object getMin() {
			return( objMin );
		}
	}
}
